KAFKA-5998: fix checkpointableOffsets handling

vvcephei opened the pull request:
* fix checkpoint file warning by filtering checkpointable offsets per task
* clean up state manager hierarchy to prevent similar bugs

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei committed:
KAFKA-5998: fix checkpointableOffsets handling

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:
```diff
@@ -48,30 +50,43 @@
  * of Global State Stores. There is only ever 1 instance of this class per Application Instance.
  */
-public class GlobalStateManagerImpl extends AbstractStateManager implements GlobalStateManager {
+public class GlobalStateManagerImpl implements GlobalStateManager {
```
Stop sharing mutable state between a superclass and subclass. The only reason to do it was to support the re-initialization logic, but the checkpoint map can just as easily be passed in as a parameter.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:
```diff
@@ -48,30 +50,43 @@
+        eosEnabled = StreamsConfig.EXACTLY_ONCE.equals(config.getString(StreamsConfig.PROCESSING_GUARANTEE_CONFIG));
+        baseDir = stateDirectory.globalStateDir();
+        checkpointFile = new OffsetCheckpoint(new File(baseDir, CHECKPOINT_FILE_NAME));
+        checkpointFileCache = new HashMap<>();
```
It took me a _really_ long time to decipher the actual purposes of "checkpoint" and "checkpointableOffsets". I've renamed them to "checkpointFile" and "checkpointFileCache" to be more self-documenting.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:
```diff
@@ -338,12 +357,12 @@ public void close(final boolean clean) throws IOException {
                     try {
                         entry.getValue().get().close();
-                    } catch (final Exception e) {
+                    } catch (final RuntimeException e) {
```
Since this PR is to clean up difficult-to-maintain code, I also included other cleanups, like dropping unnecessary `this` modifiers, restricting too-broad catch blocks, etc.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:
```diff
@@ -338,12 +357,12 @@ public void close(final boolean clean) throws IOException {
                                    .append(entry.getKey())
                                    .append(". Reason: ")
-                                   .append(e.toString())
+                                   .append(e)
```
unnecessary `toString`

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -57,16 +62,23 @@
 
     // TODO: this map does not work with customized grouper where multiple partitions
-    // of the same topic can be assigned to the same topic.
+    // of the same topic can be assigned to the same task.
```
pretty sure this was a typo

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -95,19 +111,29 @@ public ProcessorStateManager(final TaskId taskId,
-            // delete the checkpoint file after finish loading its stored offsets
-            checkpoint.delete();
-            checkpoint = null;
+            // with EOS enabled, there should never be a checkpoint file _during_ processing.
```
adding some clarity

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -95,19 +111,29 @@ public ProcessorStateManager(final TaskId taskId,
         log.debug("Created state store manager for task {} with the acquired state dir lock", taskId);
     }
 
+    void clearCheckpoints() throws IOException {
```
encapsulating this logic so we don't have to let other classes mutate our own state.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -126,7 +152,7 @@ public void register(final StateStore store,
 
         if (CHECKPOINT_FILE_NAME.equals(storeName)) {
-            throw new IllegalArgumentException(String.format("%sIllegal store name: %s", logPrefix, CHECKPOINT_FILE_NAME));
+            throw new IllegalArgumentException(String.format("%sIllegal store name: %s", logPrefix, storeName));
```
It's equivalent, but this makes more sense.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -135,10 +161,7 @@ public void register(final StateStore store,
-            registeredStores.put(storeName, Optional.of(store));
-        } else {
-
+        if (topic != null) {
```
this block has nothing to do with adding the store to `registeredStores`, so I moved that operation to after the block for clarity.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -294,11 +322,10 @@ public void close(final boolean clean) throws ProcessorStateException {
         }
 
-        if (!clean && eosEnabled && checkpoint != null) {
+        if (!clean && eosEnabled) {
```
not necessary to check whether there is a checkpoint file, since it's now encapsulated in `clearCheckpoints`.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -309,47 +336,37 @@ public void close(final boolean clean) throws ProcessorStateException {
-            } else {
-                throw new IllegalStateException("Expected " + entry.getKey() + " to have been initialized");
+    public void checkpoint(final Map<TopicPartition, Long> checkpointableOffsetsFromProcessing) {
+        ensureStoresRegistered();
```
The complicated for loop in the old version has the side-effect of checking this condition, so I'm keeping here, but making it explicit.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -309,47 +336,37 @@ public void close(final boolean clean) throws ProcessorStateException {
+    public void checkpoint(final Map<TopicPartition, Long> checkpointableOffsetsFromProcessing) {
+        ensureStoresRegistered();
+
+        final Map<TopicPartition, Long> restoredOffsets = validCheckpointableOffsets(changelogReader.restoredOffsets());
```
This was the cause of the bug. The `changelogReader` is shared by all tasks in the thread, and it returns `restoredOffsets` for _all_ tasks in the thread. So now, we filter out only the ones we actually care about.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -309,47 +336,37 @@ public void close(final boolean clean) throws ProcessorStateException {
+
+        final Map<TopicPartition, Long> restoredOffsets = validCheckpointableOffsets(changelogReader.restoredOffsets());
+        log.trace("Checkpointable offsets updated with restored offsets: {}", checkpointFileCache);
+        validCheckpointableTopics().forEachOrdered(topicPartition -> {
```
Factored out the notion of the topics that are actually checkpointable, which lets the code here just focus on picking which exact offset to use. This cuts the cyclomatic complexity of this method in half.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -380,4 +397,28 @@ void ensureStoresRegistered() {
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .filter(store -> store.persistent() && storeToChangelogTopic.containsKey(store.name()))
+            .map(store -> new TopicPartition(storeToChangelogTopic.get(store.name()), getPartition(store.name())));
```
Not sure if this is really necessary or not, but it more readable to express this logic functionally.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java:
```diff
@@ -38,45 +38,40 @@
 import static org.apache.kafka.streams.state.internals.WrappedStateStore.isTimestamped;
 
-abstract class AbstractStateManager implements StateManager {
+final class StateManagerUtil {
```
This has changed from an abstract class to a static utility class.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -286,13 +286,10 @@ public void resume() {
-                    throw new ProcessorStateException(String.format("%sError while deleting the checkpoint file", logPrefix), e);
-                }
+            try {
+                stateMgr.clearCheckpoints();
```
checking for null is now encapsulated.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -431,7 +432,7 @@ public void shouldNotChangeOffsetsIfAckedOffsetsIsNull() throws IOException {
             false,
             stateDirectory,
-            emptyMap(),
+            singletonMap(persistentStoreName, persistentStorePartition.topic()),
```
This test erroneously didn't include a changelog topic for the store in question. Now that we are actually verifying the checkpoints before we write them, we have to get this right.

vvcephei replied:
Edit: some refactoring I did actually removed this enforcement. I'm working out how to keep it, but it's pretty complicated...

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -752,7 +842,7 @@ public void shouldDeleteCheckpointFileOnCreationIfEosEnabled() throws IOExceptio
                 false,
                 stateDirectory,
-                emptyMap(),
+                singletonMap(persistentStoreName, persistentStoreTopicName),
```
ditto here.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -1477,6 +1477,8 @@ public void punctuate(final long timestamp) {
     }
 
     private StreamTask createStatefulTask(final StreamsConfig config, final boolean logged) {
+        final StateStore stateStore = new MockKeyValueStore(storeName, logged);
```
We're no longer overlooking the fact that this store wasn't "logged" when we write the checkpoints.

vvcephei committed:
avoid using potentially invalid checkpoints

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -33,15 +33,20 @@
 
 
-public class ProcessorStateManager extends AbstractStateManager {
+public class ProcessorStateManager implements StateManager {
```
Also here, no longer sharing mutable state between super and sub classes.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -57,16 +62,23 @@
+    private final File baseDir;
+    private OffsetCheckpoint checkpointFile;
+    private final Map<TopicPartition, Long> checkpointFileCache = new HashMap<>();
+    private final Map<TopicPartition, Long> initialLoadedCheckpoints;
```
Adding this collection breaks a circular dependency in this class:
* the checkpoints we load from disk are potentially not valid for the current topology
* we have to load the checkpoints immediately because we have to delete the checkpoint file before processing in the case of EOS
* we also need to have read the checkpoint file _before_ registering stores, since it might be needed to create a restorer
* we can't know if a checkpoint from the file is valid until _after_ registering stores 

In other words, if the prior code wanted to validate the loaded checkpoints, it would have to register the stores before loading checkpoints, but it also needs to load the checkpoints before registering the stores.

We're breaking the cycle here by keeping the loaded checkpoints separate. Now we read the checkpoint file into `initialLoadedCheckpoints`, which is used to register the stores, and then we are able to make sure that we only ever write valid checkpoints into the `checkpointFileCache`, which is used to update the checkpoint file later on.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -149,12 +160,16 @@ public void register(final StateStore store,
                 recordConverters.put(topic, recordConverter);
             } else {
-                log.trace("Restoring state store {} from changelog topic {} at checkpoint {}", storeName, topic, checkpointableOffsets.get(storePartition));
+                final Long restoreCheckpoint = store.persistent() ? initialLoadedCheckpoints.get(storePartition) : null;
```
This is where we're using the _loaded_ checkpoint for store registration. Note the missing condition which is now handled... if the store is not persistent, it should _not_ use the loaded checkpoint.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -164,32 +179,46 @@ public void register(final StateStore store,
+        );
+    }
+
+    void clearCheckpoints() throws IOException {
```
encapsulating this operation so that outside classes don't have to directly mutate our `checkpointFile` field.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -164,32 +179,46 @@ public void register(final StateStore store,
+            checkpointFile.delete();
+            checkpointFile = null;
+
+            checkpointFileCache.clear();
```
We didn't previously clear the cache on the blocks that this method replaces, but after reading the code, I'm pretty sure this is the right thing to do.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/OffsetCheckpoint.java:
```diff
@@ -90,6 +94,7 @@ public void write(final Map<TopicPartition, Long> offsets) throws IOException {
                 fileOutputStream.getFD().sync();
             }
 
+            LOG.trace("Swapping tmp checkpoint file {} {}", temp.toPath(), file.toPath());
```
Having these logs would have demystified a large part of the prior (misdirected) investigation, since we were never sure whether the tmp file existed or not, or what was going on.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/MockChangelogReader.java:
```diff
@@ -39,7 +40,11 @@ public void register(final StateRestorer restorer) {
+        return restoredOffsets;
+    }
+
+    void setRestoredOffsets(final Map<TopicPartition, Long> restoredOffsets) {
```
Added to support some of the needed test changes.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -240,7 +240,7 @@ public void testRegisterNonPersistentStore() throws IOException {
     public void testChangeLogOffsets() throws IOException {
         final TaskId taskId = new TaskId(0, 0);
-        final long lastCheckpointedOffset = 10L;
+        final long storeTopic1LoadedCheckpoint = 10L;
```
renamed for clarity.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -441,6 +441,123 @@ public void shouldNotChangeOffsetsIfAckedOffsetsIsNull() throws IOException {
     }
 
+    @Test
+    public void shouldIgnoreIrrlevantLoadedCheckpoints() throws IOException {
```
Added a bunch of new tests to cover both the bug itself and also previously untested code paths in the ProcessorStateManager.

vvcephei replied:
Note: this particular test shows that we will actually repair all the corrupted checkpoint files that buggy Streams versions wrote.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -441,6 +441,123 @@ public void shouldNotChangeOffsetsIfAckedOffsetsIsNull() throws IOException {
+    }
+
+    @Test
+    public void shouldIgnoreIrrelevantRestoredCheckpoints() throws IOException {
```
Note: this particular test verifies that the bug is fixed.

vvcephei reviewed (commented):
@mjsax @cadonna @ableegoldman @abbccdda  @guozhangwang @pkleindl ... did I miss anyone?

This PR should fix the long-running KAFKA-5998 bug.

The change I'm proposing is bigger in scope than the actual fix, though, because I wanted to take steps to prevent a similar bug from cropping up in the same code in the future. My theory is that the bug had an easy time hiding in this code because the handling of checkpointable offsets was so complex. I'm hoping that by reducing the mutable scope and also tightening up the invariants around the checkpointable offsets, we will have an easier time maintaining this module.

Let me know what you think!

Thanks,
-John

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -380,4 +416,33 @@ void ensureStoresRegistered() {
+        // it's only valid to record checkpoints for registered stores that are both persistent and change-logged
+
+        final Set<TopicPartition> result = new HashSet<>(storeToChangelogTopic.size());
+        for (final Map.Entry<String, String> storeToChangelog : storeToChangelogTopic.entrySet()) {
```
I would assign `storeToChangelog.getKey()` to a variable called `storeName` to make the code more readable.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -380,4 +416,33 @@ void ensureStoresRegistered() {
+
+        final Map<TopicPartition, Long> result = new HashMap<>(checkpointableOffsets.size());
+
+        for (final Map.Entry<TopicPartition, Long> entry : checkpointableOffsets.entrySet()) {
```
Could you give variable `entry` a more meaningful name?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -309,44 +338,51 @@ public void close(final boolean clean) throws ProcessorStateException {
+        }
+    }
+
+    private void updateCheckpointFileCache(final Map<TopicPartition, Long> checkpointableOffsetsFromProcessing) {
```
In this method, you call `validCheckpointableTopics` three times, once directly and twice within `validCheckpointableOffsets`. Wouldn't it make sense to call it once in the beginning and passing the result into `validCheckpointableOffsets`?

vvcephei replied:
yep. done.

vvcephei replied:
thanks for catching this.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -254,8 +254,9 @@ public void testChangeLogOffsets() throws IOException {
-        final OffsetCheckpoint checkpoint = new OffsetCheckpoint(new File(stateDirectory.directoryForTask(taskId), ProcessorStateManager.CHECKPOINT_FILE_NAME));
-        checkpoint.write(singletonMap(new TopicPartition(storeTopicName1, 0), lastCheckpointedOffset));
+        final OffsetCheckpoint checkpoint = new OffsetCheckpoint(new File(stateDirectory.directoryForTask(taskId),
+                                                                          StateManagerUtil.CHECKPOINT_FILE_NAME));
```
```suggestion
final OffsetCheckpoint checkpoint = new OffsetCheckpoint(
    new File(stateDirectory.directoryForTask(taskId), StateManagerUtil.CHECKPOINT_FILE_NAME));

```

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -441,6 +441,123 @@ public void shouldNotChangeOffsetsIfAckedOffsetsIsNull() throws IOException {
     }
 
+    @Test
+    public void shouldIgnoreIrrlevantLoadedCheckpoints() throws IOException {
```
```suggestion
    public void shouldIgnoreIrrelevantLoadedCheckpoints() throws IOException {
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -33,15 +33,20 @@
 
 
-public class ProcessorStateManager extends AbstractStateManager {
+public class ProcessorStateManager implements StateManager {
```
Wouldn't it be more meaningful to rename this class to `TaskStateManager`?

vvcephei replied:
Maybe, I'm not sure of the historical reason to name it this way.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -309,44 +338,51 @@ public void close(final boolean clean) throws ProcessorStateException {
+    }
+
+    private void updateCheckpointFileCache(final Map<TopicPartition, Long> checkpointableOffsetsFromProcessing) {
+        final Map<TopicPartition, Long> restoredOffsets = validCheckpointableOffsets(changelogReader.restoredOffsets());
```
This is the most important line to fix the bug, right?

vvcephei replied:
Yes. I'd marked it in an earlier version of this PR. I guess that comment became "outdated" at some point. Sorry about that.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -149,12 +160,16 @@ public void register(final StateStore store,
                     storePartition,
                     new CompositeRestoreListener(stateRestoreCallback),
-                    checkpointableOffsets.get(storePartition),
+                    restoreCheckpoint,
```
If the store is not peristent or the read checkpoint file does not contain the partition, this will throw a NPE, right? If yes, you should add unit tests for these cases.

vvcephei replied:
No, I guess you were thinking the `Long` would become unboxed at this point? It's actually a `Long` parameter, and the `StateRestorer` constructor checks for null... Not the cleanest code, I guess, but it looks like it's been this way since 2017.

cadonna replied:
My fault! I missed the parameter. I looked at the next parameter in the `StateRestorer` constructor which is a `long`.

cadonna reviewed (commented):
Thanks for the PR @vvcephei and congratulations to you and @pkleindl for finding and fixing this bug. I left a couple of comments. I had hard time to find the code that actually fixes the bug (and I am still not sure if I found it). Could you please add some specific comments about the fix, next time, since this fix is not that trivial? I am also wondering if you could have divided this PR in two: one for the fix itself and one for the repair of old corrupted checkpoints. IMO, it would have made reviewing the PRs easier.

vvcephei committed:
cr comments

vvcephei commented:
Thanks for the review, @cadonna !

I'm sorry that my earlier call-out of the actual bugfix got marked "outdated" at some point, so I guess it doesn't show up in the diff anymore. I didn't notice when that happened, or I would have re-marked it.

Regarding splitting up the PRs, I do agree with you. It would have been nice to get a smaller fix in, and then tackled the refactoring separately.

If I can make one excuse for myself, it would be that in this case, it wasn't clear to me that the fix was good enough because the scope of `checkpointableOffsets` was too broad. If we just did a bugfix first, it would have been equally hard for reviewers to have confidence in the fix.

In retrospect, though, I could have submitted the refactor first, and then followed up with the bugfix. It just didn't occur to me, for whatever reason.

In any case, thanks for wading through the code review! I think I addressed all your comments.

WDYT?
-John

bbejeck commented:
retest this please

cadonna reviewed (approved):
LGTM

bbejeck commented:
Merged #7030 into trunk

bbejeck commented:
cherry-picked to 2.3 and 2.2

Tin-Nguyen commented:
@bbejeck  it means that the fix is available in 2.2.x version?

bbejeck commented:
@Tin-Nguyen yes if check out the `2.2` branch and build it.  For right now I don't know if there will be another `2.2` release.

Tin-Nguyen commented:
thanks @bbejeck

Tin-Nguyen commented:
@bbejeck  I'm wondering if we have an updated binary download includes the fix?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java:
```diff
@@ -38,45 +38,40 @@
-        this.eosEnabled = eosEnabled;
-        this.checkpoint = new OffsetCheckpoint(new File(baseDir, CHECKPOINT_FILE_NAME));
-    }
+    private StateManagerUtil() {}
```
Do we need this constructor explicitly? Would this just be default in java?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -164,32 +179,46 @@ public void register(final StateStore store,
-            storeToChangelogTopic,
-            partitions,
-            processorContext);
+        StateManagerUtil.reinitializeStateStoresForPartitions(log,
```
I liked this refactoring a lot, thanks @vvcephei !
